### PR TITLE
Tanks in series

### PR DESCRIPTION
**Pull request by parkec3**
# Tanks-In-Series Model

Implementation of the tanks in series model within Pybamm. Based on DOI: 10.1149/2.0342001JES. 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by parkec3** [comment]
I have the model written up in a .py file, but I think there is an issue with domains/discretization. I'm getting this error when I run the following code:

image

image

Initially, I got an error of size (400,1) vs (1,1), and then I used `pybamm.x_average(param.n.prim.c_init)`, but the size still doesn't match up.

**Comment by valentinsulzer** [comment]
You need to average in both x- and r-directions, you can use `param.n.prim.c_init_av` which is already averaged

File: pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+        c_surf_ave_n = pybamm.Variable(
+            "X-averaged negative particle surface concentration [mol.m-3]"
+        )
+        c_surf_ave_p = pybamm.Variable(
+            "X-averaged positive particle surface concentration [mol.m-3]"
+        )
+
+        q_ave_n = pybamm.Variable("X-averaged negative particle concentration gradient")
+        q_ave_p = pybamm.Variable("X-averaged positive particle concentration gradient")
```
Comment (valentinsulzer) [inline pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py:51]: can you add units to these?

File: pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+        self.rhs[c_s_n] = -3 * j_n / param.n.prim.R_typ
+        self.rhs[c_s_p] = -3 * j_p / param.p.prim.R_typ
+
+        self.initial_conditions[c_s_n] = pybamm.x_average(param.n.prim.c_init)
+        self.initial_conditions[c_s_p] = pybamm.x_average(param.p.prim.c_init)
+
+        # Discretized Particle Concentration Gradient
+        q_ave_n = pybamm.Variable("X-averaged negative particle concentration gradient")
+        q_ave_p = pybamm.Variable("X-averaged positive particle concentration gradient")
```
Comment (valentinsulzer) [inline pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py:118]: Looks like these are duplicate

File: pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            + 4
+            * param.R
+            * T
+            * param.F
+            * param.chi(c_12, T)
+            / 2
+            * param.kappa_e(c_12, T)
+            / c_12
+            * (c_e_s - c_e_p)
+            / leps12
```
Comment (valentinsulzer) [inline pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py:216]: grouping things in brackets, or precomputing some things (e.g. `RT_F = param.R * T / param.F`) will make this easier to read

File: pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        # Current in the electrolyte
+        ######################
+
+        self.algebraic[phi_e_p] = (
+            -iapp
+            - 2 * param.kappa_e(c_12, T) * (phi_e_s - phi_e_p) / leps12
+            + 4
+            * param.R
+            * T
+            * param.F
```
Comment (valentinsulzer) [inline pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py:210]: should this be `/ param.F` ?

**Review by valentinsulzer — COMMENTED** [review]
Definitely on the right track! Still needs tests and documentations, we can talk about how to add those.

File: pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py (hunk 4 above)
Comment (parkec3) [reply pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py:210]: Yes, fixed it.

File: pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py (hunk 1 above)
Comment (parkec3) [reply pybamm/models/full_battery_models/lithium_ion/tanks_in_series.py:51]: Units added.

**Comment by parkec3** [comment]
It seems to be working now with Chen2020 parameter set.

**Comment by rtimms** [comment]
Looking good! Have you done a comparison to other models, e.g. using [this example](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/scripts/compare_lithium_ion.py)?

**Pull request closed without merging**